### PR TITLE
Implemented an owned Consumer api for OrdZSet

### PR DESCRIPTION
**Pull request by Kixiron**
Found a fairly basic consumer api that seems to work, only implemented it for `OrdZSet` for now

### CODE REVIEW

File: src/trace/cursor/mod.rs [hunk 1]
```diff
@@ -113,6 +113,28 @@ pub trait Cursor<'s, K, V, T, R> {
...
+
+    fn next_key(&mut self) -> (K, Self::ValueConsumer<'_>);
+
+    fn seek_key(&mut self, key: &K)
+    where
+        K: Ord;
+}
+
+pub trait ValueConsumer<'a, V, R> {
+    fn value_valid(&self) -> bool;
```
Comment (ryzhyk) [inline src/trace/cursor/mod.rs:131]: Let's call this `val_valid` for consistency with `Cursor`.

File: src/trace/cursor/mod.rs [hunk 2]
```diff
@@ -113,6 +113,28 @@ pub trait Cursor<'s, K, V, T, R> {
     fn rewind_vals(&mut self);
 }
 
+pub trait Consumer<K, V, R> {
```
Comment (ryzhyk) [inline src/trace/cursor/mod.rs:116]: Can you please document the new traits?

**Review by ryzhyk — APPROVED** [review]

File: src/trace/cursor/mod.rs (hunk 1 above)
Comment (Kixiron) [reply src/trace/cursor/mod.rs:140]: I personally dislike the shortened names to be honest, that's why I changed it

**Comment by Kixiron** [comment]
Extended this to work with `OrdIndexedZSet` and `OrderedLayer<K, OrderedColumnLeaf<V, R>, O>`

**Comment by Kixiron** [comment]
Implemented consumers for `OrdKeyBatch`

File: src/operator/aggregate/average.rs [hunk 3]
```diff
@@ -153,11 +154,58 @@ where
...
             self.aggregate_linear::<TS, _, _>(move |key, val| Avg::new(f(key, val), 1isize));
 
-        // TODO: We can probably use some sort of `.map_index_owned()` here since in all
-        // likelihood we'll be the only consumer of the aggregated value
-        // (meaning we wouldn't need to clone the average's sum and would be
-        // able to elide one clone of the key value)
-        let average = aggregate.map_index(|(k, avg)| (k.clone(), (avg.sum.clone()) / avg.count));
+        // We're the only possible consumer of the aggregated stream so we can use an
+        // owned consumer to skip any cloning
+        let average = aggregate.apply_owned_named("MapAverage", |aggregate| {
```
Comment (ryzhyk) [inline src/operator/aggregate/average.rs:159]: Any chance you could add a test (proptest?) for the `average` operator?

Comment (ryzhyk) [reply src/operator/aggregate/average.rs:159]: Also, how does this affect the performance of q4, which is the only benchmark that uses `average` at the moment?

Comment (Kixiron) [reply src/operator/aggregate/average.rs:159]: On my machine with 1 thread & 1 generator I get 89.412s for the previous impl and 74.469s with my change

**Pull request merged**